File: src/main/java/dev/kensa/sentence/Dictionary.java
package dev.kensa.sentence;

import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;

import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toSet;

public final class Dictionary {

    private final Set<Acronym> acronyms = new LinkedHashSet<>();
    private final Set<String> keywords = new LinkedHashSet<>(asList("given", "when", "then", "and", "with", "that"));

    public void putAcronyms(Acronym... acronyms) {
        Objects.requireNonNull(acronyms);

        this.acronyms.addAll(asList(acronyms));
    }

    public void putKeyword(String value) {
        if (value == null || value.length() < 2) {
            throw new IllegalArgumentException(String.format("Keywords must be at least 2 characters. [%s]", value));
        }
        keywords.add(value);
    }

    public void clearAcronyms() {
        acronyms.clear();
    }

    public void putKeywords(String... values) {
        asList(values).forEach(this::putKeyword);
    }

    public Set<Acronym> acronyms() {
        return acronyms;
    }

    public Set<String> keywords() {
        return keywords;
    }

    public Set<String> acronymStrings() {
        return acronyms.stream().map(Acronym::acronym).collect(toSet());
    }
}

File: src/main/java/dev/kensa/sentence/Dictionary.kt
package dev.kensa.sentence

class Dictionary {
    private val highlightedIdentifiers: MutableSet<HighlightedIdentifier> = LinkedHashSet()

    fun putHighlightedIdentifiers(vararg identifiers: HighlightedIdentifier) {
        highlightedIdentifiers.addAll(identifiers)
    }

    fun putHighlightedIdentifiers(identifiers: Set<HighlightedIdentifier>) {
        highlightedIdentifiers.addAll(identifiers)
    }

    private val _acronyms: MutableSet<Acronym> = LinkedHashSet()
    val acronyms: Set<Acronym>
        get() = _acronyms

    fun putAcronyms(vararg acronyms: Acronym) {
        _acronyms.addAll(acronyms)
    }

    fun putAcronyms(acronyms: Set<Acronym>) {
        _acronyms.addAll(acronyms)
    }

    fun clearAcronyms() {
        _acronyms.clear()
    }

    private val _keywords: MutableSet<Keyword> = linkedSetOf(
        Keyword("given"),
        Keyword("when"),
        Keyword("whenever"),
        Keyword("then"),
        Keyword("and"),
        Keyword("with"),
        Keyword("that")
    )
    val keywords: Set<Keyword>
        get() = _keywords

    fun putKeyword(keyWord: Keyword) {
        _keywords.add(keyWord)
    }

    fun putKeyword(value: String) {
        _keywords.add(Keyword(value))
    }

    fun putKeywords(vararg values: String) {
        values.forEach(this::putKeyword)
    }

    fun putKeywords(vararg keywords: Keyword) {
        keywords.forEach(this::putKeyword)
    }

    fun putKeywords(values: Set<String>) {
        values.forEach(this::putKeyword)
    }

    fun findInterestingIdentifierOrNull(value: String) = highlightedIdentifiers.firstOrNull { it.value == value }
    fun isAcronym(value: String) = acronyms.any { it.acronym.equals(value, ignoreCase = true) }
    fun findKeywordOrNull(value: String) = keywords.firstOrNull { it.value == value }
    fun isWhen(value: String) = value.equals("when", ignoreCase = true) || value.equals("whenever", ignoreCase = true)
}